napi-inl.h

```diff
@@ -3370,9 +3370,32 @@ inline napi_value ObjectWrap<T>::ConstructorCallbackWrapper(
   }
 
   T* instance;
+  auto removeFailedWrap = [](const CallbackInfo& info) {
+    napi_status status = napi_remove_wrap(info.Env(), info.This(), nullptr);
+    NAPI_FATAL_IF_FAILED(status,
+        "ObjectWrap<T>::ConstructorCallbackWrapper",
+        "Failed to remove wrap from failed ObjectWrap instance construction");
+  };
   napi_value wrapper = details::WrapCallback([&] {
     CallbackInfo callbackInfo(env, info);
+#ifdef NAPI_CPP_EXCEPTIONS
+    try {
+      instance = new T(callbackInfo);
+    } catch (const Error& e) {
+      removeFailedWrap(callbackInfo);
+      // re-throw the error after removing the failed wrap.
+      throw e;
+    }
+#else
     instance = new T(callbackInfo);
+    if (callbackInfo.Env().IsExceptionPending()) {
+      // We need to clear the exception so that `napi_remove_wrap()` might work.
+      Error e = callbackInfo.Env().GetAndClearPendingException();
+      removeFailedWrap(callbackInfo);
+      e.ThrowAsJavaScriptException();
+      delete instance;
+    }
+# endif  // NAPI_CPP_EXCEPTIONS
```

Comment:
This assumes that `ObjectWrap<T>::ObjectWrap()` succeeded, but the subclass constructor failed. If `ObjectWrap<T>::ObjectWrap()` itself fails, for example, if `napi_wrap()` returns non-`napi_ok`, then no wrapping will have taken place, and `napi_remove_wrap()` will fail. In this latter case `napi_remove_wrap()` should not be attempted, but rather, the existing error should be surfaced. This means we have to distinguish between three scenarios:

0. The constructor succeeded.
0. `napi_wrap()` succeeded, but the subclass constructor threw an error.
0. `napi_wrap()` failed.